### ISSUE TITLE
Generated GML files miss the CRS

### ISSUE THREAD

**Issue by pathmapper** [qgis/QGIS#43725]
**Describe the bug**
processing and export - GML output misses CRS

**How to Reproduce**

1. Add [test_data_gml_output.zip](https://github.com/qgis/QGIS/files/6648955/test_data_gml_output.zip) to a QGIS project 
2. _Export_ -> _Save features as ..._ -> GML output with CRS specified
<img src="https://user-images.githubusercontent.com/20856381/121908241-171e0d80-cd2d-11eb-8e37-e87086b4ac81.png" width="500">

3. Use the _Dissolve_ processing tool with GML output
<img src="https://user-images.githubusercontent.com/20856381/121907888-c73f4680-cd2c-11eb-9867-46c508f6da91.png" width="500">

4. See error: No CRS specified for GML output / no `.gfs` is created

<img src="https://user-images.githubusercontent.com/20856381/121908374-3c128080-cd2d-11eb-95dd-7567be49a8a1.png" width="200">

**QGIS and OS versions**

QGIS version | 3.19.0-Master | QGIS code revision | b3fa3474
-- | -- | -- | --
Qt version | 5.15.2
Python version | 3.9.5
GDAL/OGR version | 3.2.2
PROJ version | 8.0.1
EPSG Registry database version | v10.018 (2021-04-02)
GEOS version | 3.9.1-CAPI-1.14.2
SQLite version | 3.35.5
PostgreSQL client version | 11.5
SpatiaLite version | 5.0.1
QWT version | 6.1.5
QScintilla2 version | 2.11.5
OS version | Windows 10 Version 1909
&nbsp; | &nbsp; | &nbsp; | &nbsp;
This copy of QGIS writes debugging output.
&nbsp; | &nbsp; | &nbsp; | &nbsp;
Active Python plugins | canvas_clipperLAStoolsplugin_reloaderdb_managerMetaSearchprocessing

**Additional context**

Regarding the processing GML output, there's this resolved issue: https://github.com/qgis/QGIS/issues/22517

**Comment by gioman** [comment]
@pathmapper not sure if this the explanation, but...

also doing a "export as..." to a GML creates a file that QGIS will not recognize its CRS.

In GML the CRS is defined per geometry, not a layer level, in fact if you ogrinfo a GML you get (example)


giovanni@***:~$ ogrinfo -so t3.gml t3
INFO: Open of `t3.gml'  using driver `GML' successful.
Layer name: t3
Geometry: Multi Polygon
Feature Count: 1
Extent: (5.945017, 51.166833) - (6.940634, 51.905244)
**Layer SRS WKT:
(unknown)**
Geometry Column = geometryProperty
fid: String (0.0) NOT NULL
BUNDESLAND: String (0.0)
BUNDESLANDSCHLUESSEL: String (0.0)
GDO_GID: Integer (10.0)
KREIS: String (0.0)
KREISSCHLUESSEL: Integer (10.0)

**Comment by pathmapper** [comment]
@gioman 

> "export as..." to a GML creates a file that QGIS will not recognize its CRS.

Yes, I'm aware of this (point 2. from "How to reproduce" above).

After reading the mentioned ticket https://github.com/qgis/QGIS/issues/22517 I've got the impression that this is a regression.

From https://github.com/qgis/QGIS/issues/22517#issuecomment-495858308:

> The "save as..." works properly, creating a .gml and a .gfs file, with the correct CRS.

Currently no `.gfs` file is created.

And here's the related PR from 5 years ago: https://github.com/qgis/QGIS/pull/3381

From https://github.com/qgis/QGIS/pull/3381#issue-80925010:

> Fixes missing CRS when saving to GML

The latest GML spec is from 2012 (https://www.ogc.org/standards/gml), so there should be no changes to the spec since this issue was fixed in QGIS in 2016.

**Comment by gioman** [comment]
So rather than Processing issue this seems a more general one. Tagging this as a regression waiting for a more authoritative feedback...

**Comment by pathmapper** [comment]
> rather than Processing issue this seems a more general one

Yep, I think so, too. Sorry @gioman if this wasn't clear from the issue description.

[PolygonDissolveTest_output.zip](https://github.com/qgis/QGIS/files/6650349/PolygonDissolveTest_output.zip) contains a `.gml` file with related `.gfs` file from  
https://github.com/qgis/QGIS/tree/master/python/plugins/processing/tests/testdata/expected

I wonder why there is `.gfs` test data but no failing test related to this.

```
ogrinfo -so PolygonDissolveTest_output.gml PolygonDissolveTest_output
INFO: Open of `PolygonDissolveTest_output.gml'
      using driver `GML' successful.

Layer name: PolygonDissolveTest_output
Geometry: Multi Polygon
Feature Count: 1
Extent: (1754260.455660, 5078929.555540) - (1757399.087430, 5080666.093610)
Layer SRS WKT:
PROJCS["Monte Mario / Italy zone 1",
    GEOGCS["Monte Mario",
        DATUM["Monte_Mario",
            SPHEROID["International 1924",6378388,297,
                AUTHORITY["EPSG","7022"]],
            TOWGS84[-104.1,-49.1,-9.9,0.971,-2.917,0.714,-11.68],
            AUTHORITY["EPSG","6265"]],
        PRIMEM["Greenwich",0,
            AUTHORITY["EPSG","8901"]],
        UNIT["degree",0.0174532925199433,
            AUTHORITY["EPSG","9122"]],
        AUTHORITY["EPSG","4265"]],
    PROJECTION["Transverse_Mercator"],
    PARAMETER["latitude_of_origin",0],
    PARAMETER["central_meridian",9],
    PARAMETER["scale_factor",0.9996],
    PARAMETER["false_easting",1500000],
    PARAMETER["false_northing",0],
    UNIT["metre",1,
        AUTHORITY["EPSG","9001"]],
    AXIS["X",EAST],
    AXIS["Y",NORTH],
    AUTHORITY["EPSG","3003"]]
fid: String (0.0) NOT NULL
```

**Comment by nyalldawson** [comment]
This is an upstream (GDAL) related change.

**Comment by pathmapper** [comment]
Export GPKG as GML results in `.gml` and `.xsd` files:

![grafik](https://user-images.githubusercontent.com/20856381/136895702-1842b0a0-033f-464f-af00-d1deb6ebbc49.png)

When the `.gml` file is opened in QGIS, the CRS is not detected: 

![grafik](https://user-images.githubusercontent.com/20856381/136895884-d406bb6e-f71c-4372-b9c7-9ff15e9d470c.png)

---

If the `.xsd` file is deleted:

![grafik](https://user-images.githubusercontent.com/20856381/136896117-b5f65236-a916-4e8b-8789-8a6d40d3a8e2.png)

When the `.gml` file without `.xsd` is opened in QGIS, the CRS is detected:


![grafik](https://user-images.githubusercontent.com/20856381/136896389-9d01b259-b195-4fcb-b38a-0a3f1b704719.png)

and a `.gfs` was created by opening the `.gml` file without `.xsd` in QGIS:


![grafik](https://user-images.githubusercontent.com/20856381/136896526-8a9923d1-4a4e-4ea3-9eb8-a0854284900e.png)

---

After reading https://gdal.org/drivers/vector/gml.html, I've got the impression that maybe the issue (CRS of `.gml` is not detected) could be solved by using one of the following two open options:

https://gdal.org/drivers/vector/gml.html#open-options 

**WRITE_GFS=YES** (so that a GFS file is always created and QGIS is able to detect the CRS)

> WRITE_GFS=AUTO/YES/NO: (GDAL >= 3.2) whether to write a .gfs file. In AUTO mode, the .gfs file is only written if there is no recognized .xsd file, no existing .gfs file and for non-network file systems. This option can be set to YES for force .gfs file writing in situations where AUTO would not attempt to do it. Or it can be set to NO to disable .gfs file writing.

or

**FORCE_SRS_DETECTION=YES** (so that QGIS detects the CRS from `.xsd`, no `.gfs` needed)

> FORCE_SRS_DETECTION=YES/NO: Force a full scan to detect the SRS of layers. This option may be needed in the case where the .gml file is accompanied with a .xsd. Normally in that situation, OGR would not detect the SRS, because this requires to do a full scan of the file. Defaults to NO

**Comment by pathmapper** [comment]
Maybe we should use `FORCE_SRS_DETECTION=YES` if no `.gfs` file is available.

**Comment by agiudiceandrea** [comment]
@pathmapper I can confirm your findings on Windows using QGIS 3.16.16 from OSGeo4W v1 (GDAL 3.1.4).

The issue seems no longer occurring on Windows using QGIS 3.16.16 and QGIS 3.20.3 both from OSGeo4W v2 (GDAL 3.4.1): exporting in GML format the .gml and .xsd files are created, however the layer CRS is correctly recognized when such layer is imported in QGIS. The layer CRS is correctly recognized even deleting the .xsd file (in this case a .gfs file is automatically generated when the layer is imported in QGIS).

**Comment by gioman** [comment]
> The issue seems no longer occurring on Windows using QGIS 3.16.16 and QGIS 3.20.3 both from OSGeo4W v2 (GDAL 3.4.1): exporting in GML format the .gml and .xsd files are created, however the layer CRS is correctly recognized when such layer is imported in QGIS. The layer CRS is correctly recognized even deleting the .xsd file (in this case a .gfs file is automatically generated when the layer is imported in QGIS).

Then I think we can move on and close this ticket.

**Comment by pathmapper** [comment]
Thanks @agiudiceandrea 👍 

Confirmed this is fixed with 3.22.3 and GDAL 3.4.1 on Windows.
